numpy/core/setup_common.py

```diff
@@ -120,15 +120,29 @@ def set_sig(sig):
             set_sig(line)
 
 # Mandatory functions: if not found, fail the build
-MANDATORY_FUNCS = ["sin", "cos", "tan", "sinh", "cosh", "tanh", "fabs",
-        "floor", "ceil", "sqrt", "log10", "log", "exp", "asin",
-        "acos", "atan", "fmod", 'modf', 'frexp', 'ldexp']
+MANDATORY_FUNCS = [
+    "sin", "cos", "tan", "sinh", "cosh", "tanh", "fabs",
+    "floor", "ceil", "sqrt", "log10", "log", "exp", "asin",
+    "acos", "atan", "fmod", 'modf', 'frexp', 'ldexp',
+    "expm1", "log1p", "acosh", "asinh", "atanh",
+    "rint", "trunc", "exp2", 
+    "copysign", "nextafter", "strtoll", "strtoull", "cbrt",
+]
+
+OPTIONAL_STDFUNCS_BASE = [
+    # cygwin
+    "log2",
+    # macos for powl
+    "pow",
+    # 32-bit windows
+    "hypot",
+    # 32-bit mingw, visual studio 2015
+    "atan2",
+]
 
-# Standard functions which may not be available and for which we have a
-# replacement implementation. Note that some of these are C99 functions.
-OPTIONAL_STDFUNCS = ["expm1", "log1p", "acosh", "asinh", "atanh",
-        "rint", "trunc", "exp2", "log2", "hypot", "atan2", "pow",
-        "copysign", "nextafter", "strtoll", "strtoull", "cbrt"]
+OPTIONAL_STDFUNCS = OPTIONAL_STDFUNCS_BASE[:]
+OPTIONAL_STDFUNCS += [f + 'f' for f in OPTIONAL_STDFUNCS_BASE]
+OPTIONAL_STDFUNCS += [f + 'l' for f in OPTIONAL_STDFUNCS_BASE]
```

Comment:
So this logic is still a little more complicated than it should be. These still C99 functions, so wouldn't it be easier to simply say here that they are mandatory? The blocklist is not connected directly to what is happening here; there should not be a case of `pow` et al. not existing at all right?

Reply:
The name `OPTIONAL_STDFUNCS` is legacy. What about `BLOCKLISTABLE` instead?

Reply:
Better, but still the question remains why it should be a separate list. I'd say merge it all into MANDATORY, and just add a comment xref'ing the blocklist that is defined elsewhere.

Reply:
Maybe I am not expressing myself clearly enough: these functions are part of C99 and simply must exist just like all other mandatory functions. There is no way that `HAVE_POW` & co are not going to be `1`. There is no reason to treat them differently from the other C99 functions, unless I am missing something. The use of `HAVE_POW` in the blocklist can simply be removed. This cannot fail any more than `HAVE_SIN` et al.

Reply:
> There is no way that HAVE_POW & co are not going to be 1

The macro `HAVE_POW` must exist for the current blocklist mechanism to be able to undefine it. Since `sin` is not blocklist-able, `HAVE_SIN` is not necessary and is not generated in this PR. Maybe it would be more clear if the blocklist worked differently: it would define a `BLOCK_POW`. Then the code would do
```
double pow(double x, double y) {
#ifndef(BLOCK_POW)
    return pow(x, y);
#else 
    // blocklisted implementation
    ...
#endif
};
```

instead of the current
```
double pow(double x, double y) {
#ifdef(HAVE_POW)
    return pow(x, y);
#else 
    // blocklisted implementation
    ...
#endif
};
```

Reply:
Yes, that sounds nicer - shorter and clearer. It also helps keep the meson configure output concise (there'll be one line per check, so unnecesary checks are good to cut out).

Reply:
Done for the non-complex blocked functions (assuming CI is happy). The complex ones are more ... complex to convert.

Reply:
Thanks! Yeah, I think we can leave complex alone in this PR, because it's a large list.

Reply:
Eventually we'll rip this code out anyway:)